File: figures/10-state-space-search/snippet-color-graph-coloring.c
static bool safe(const unsigned char *adj, size_t n, size_t v, int color,
                 const int *colors) {
    size_t u;
    for (u = 0; u < v; u++)
        if (adj[v * n + u] && colors[u] == color) return false;
    return true;
}

static bool dfs(const unsigned char *adj, size_t n, int m, size_t v, int *colors) {
    int c;
    if (v == n) return true;
    for (c = 1; c <= m; c++) {
        if (safe(adj, n, v, c, colors)) {
            colors[v] = c;
            if (dfs(adj, n, m, v + 1, colors)) return true;
            colors[v] = 0;
        }
    }
    return false;
}

ss_status ss_color_graph(const unsigned char *adjacency, size_t n, int m,
                         int *colors) {
    size_t i;
    if (adjacency == NULL || colors == NULL || m < 0) return SS_INVALID;
    for (i = 0; i < n; i++) colors[i] = 0;
    return dfs(adjacency, n, m, 0, colors) ? SS_OK : SS_NO_SOLUTION;
}

File: figures/10-state-space-search/snippet-color-graph-coloring.java
    public int[] color(int[][] graph, int colorCount) {
        if (graph == null || colorCount < 0) throw new IllegalArgumentException();
        int n = graph.length;
        for (int[] row : graph) if (row == null || row.length != n) throw new IllegalArgumentException();
        int[] colors = new int[n];
        return dfs(graph, colorCount, 0, colors) ? colors : null;
    }

    private boolean dfs(int[][] g, int m, int v, int[] c) {
        if (v == g.length) return true;
        for (int color = 1; color <= m; color++) {
            if (safe(g, v, color, c)) {
                c[v] = color;
                if (dfs(g, m, v + 1, c)) return true;
                c[v] = 0;
            }
        }
        return false;
    }

    private boolean safe(int[][] g, int v, int color, int[] c) {
        for (int u = 0; u < v; u++) if (g[v][u] != 0 && c[u] == color) return false;
        return true;
    }
